util/src/main/java/com/coderanch/util/cli/InputUtility.java

```diff
@@ -0,0 +1,276 @@
+     */
+    public InputUtility() {
+        this(System.in, StandardCharsets.UTF_8);
+    }
+
+    /**
+     * Constructor that takes an input stream to be used.
+     * This constructor will buffer the {@code inputStream}.
+     * You cannot use the same stream in a different utility.
+     *
+     * @param inputStream input stream to be used.
+     * @param charset     charset to be used.
+     */
+    public InputUtility(InputStream inputStream, Charset charset) {
+        requireThat("inputStream", inputStream, is(notNullValue()));
+        requireThat("charset", charset, is(notNullValue()));
+        reader = new BufferedReader(new InputStreamReader(inputStream, charset));
+    }
+
+    /**
+     * Displays the prompt and reads the input stream, returning a validated String.
+     *
+     * @param prompt          the prompt to display to the user.
+     * @param stringPredicate the predicate to use for validation.
+     * @return a possibly validated String.
+     * @throws IOException when there's a problem with the underlying {@link BufferedReader}.
+     */
+    public String nextString(String prompt, Predicate<? super String> stringPredicate) throws IOException {
+        System.out.println(prompt);
+        String line;
+        for (
+                line = reader.readLine();
+                !stringPredicate.test(line);
+                line = reader.readLine()
+        ) {
+            System.out.println("Invalid input.");
+        }
+
+        return line;
+    }
+
+    /**
+     * Displays the prompt and reads the input stream, returning a validated integer.
+     *
+     * @param prompt       the prompt to display to the user.
+     * @param intPredicate the predicate to use for validation.
+     * @return a possibly validated integer.
+     * @throws IOException when there's a problem with {@link InputStream}
+     */
+    public int nextInt(String prompt, Predicate<Integer> intPredicate) throws IOException {
+        System.out.println(prompt);
+        String line;
+        Integer num;
+        for (
+                line = reader.readLine();
+                (num = this.tryIntParse(line, intPredicate)) == null;
+                line = reader.readLine()
+        ) {
+            System.out.println("Invalid input.");
+        }
+
+        return num;
+    }
+
+    private Integer tryIntParse(String line, Predicate<Integer> intPredicate) {
+        try {
+            var num = Integer.parseInt(line);
+            if (intPredicate.test(num)) {
+                return num;
+            }
+            else {
+                return null;
+            }
+        }
+        catch (Exception e) {
+            return null;
+        }
+    }
+
+    /**
+     * Displays the prompt and reads the input stream, returning a validated real number.
+     *
+     * @param prompt          the prompt to display to the user.
+     * @param doublePredicate the predicate to use for validation.
+     * @return a possibly validated real number.
```

Comment:
Doubles can't represent real numbers, only a subset of rational numbers.

```suggestion
     * @return a validated floating point number.
```